File: bench/agent-eval/fixtures/ctf-permissions/app-ts/src/routes/export.ts
// Export routes. Because the four permission abstractions were introduced in
// different releases, THIS file gates the export permission four different
// ways depending on when each handler was written.
import { User, currentUser } from "../models/user";
import { permissions } from "../perm/service";
import { requirePermission } from "../perm/guard";
import { enforceConfigRule, RuleContext } from "../perm/rules";
import { PERM_EXPORT } from "../perm/constants";

// --- release 1 handler: inline boolean flag (abstraction 1) ---
export function exportCsv(req: any, res: any): void {
  const user: User = currentUser(req);
  if (!user.canExport) {
    res.status(403).send("no export");
    return;
  }
  res.send("csv-data");
}

// --- release 1 handler: a second inline flag gate (abstraction 1) ---
export function exportPdf(req: any, res: any): void {
  const user: User = currentUser(req);
  if (!user.canExport) {
    res.status(403).send("no export");
    return;
  }
  res.send("pdf-data");
}

// --- release 2 handler: permission service (abstraction 2) ---
export function exportJson(req: any, res: any): void {
  const user: User = currentUser(req);
  permissions.require(user, PERM_EXPORT);
  res.send("json-data");
}

// --- release 2 handler: a second service gate (abstraction 2) ---
export function exportXml(req: any, res: any): void {
  const user: User = currentUser(req);
  permissions.require(user, PERM_EXPORT);
  res.send("xml-data");
}

// --- release 4 handler: config-file rule (abstraction 4) ---
export function exportArchive(req: any, res: any): void {
  const user: User = currentUser(req);
  const ctx: RuleContext = { user };
  enforceConfigRule(ctx, PERM_EXPORT);
  res.send("archive-data");
}

// --- release 3 wiring: middleware guard applied at the router (abstraction 3) ---
export function registerExportRoutes(router: any): void {
  router.get("/export/csv", exportCsv);
  router.get("/export/pdf", exportPdf);
  router.get("/export/json", exportJson);
  router.get("/export/xml", exportXml);
  router.get("/export/archive", exportArchive);
  // Bulk export was locked down later with the guard middleware.
  router.post("/export/bulk", requirePermission(PERM_EXPORT), (req: any, res: any) => {
    res.send("bulk-data");
  });
}

File: v5/bench/agent-eval/fixtures/ctf-permissions/app-rs/src/handlers/export.rs
//! Export handlers. Each was written in a different release, so each gates the
//! export permission through a different abstraction.
use crate::models::{PermResult, User};
use crate::perm_config::{enforce_rule, RuleContext, EXPORT_RULE};
use crate::perm_flags::check_export_flag;
use crate::perm_guard::require_export;
use crate::perm_service::{Permission, PermissionSet};

// release 1: raw boolean flag (abstraction 1).
pub fn export_csv(user: &User) -> PermResult {
    check_export_flag(user)?;
    Ok(())
}

// release 2: permission service (abstraction 2).
pub fn export_json(user: &User) -> PermResult {
    let perms = PermissionSet::new(user);
    perms.require(Permission::CanExport)?;
    Ok(())
}

// release 3: guard helper (abstraction 3).
pub fn export_bulk(user: &User) -> PermResult {
    require_export(user)?;
    Ok(())
}

// release 4: config-file rule (abstraction 4).
pub fn export_archive(user: &User) -> PermResult {
    let ctx = RuleContext { user };
    enforce_rule(&ctx, EXPORT_RULE)?;
    Ok(())
}
